pkg/cri/server/sandbox_run.go

```diff
@@ -218,12 +192,17 @@ func (c *criService) RunPodSandbox(ctx context.Context, r *runtime.RunPodSandbox
 		containerd.WithContainerExtension(sandboxMetadataExtension, &sandbox.Metadata),
 		containerd.WithRuntime(ociRuntime.Type, runtimeOpts)}
 
+	// A flag to keep the container in metadata store even in error case.
+	// Combined with adding sandbox to sandboxstore, it allows kubelet to reconcile the sandbox status for cleanup.
+	// e.g. when teardownPodNetwork fails
+	shouldDelContainer := true
+
 	container, err := c.client.NewContainer(ctx, id, opts...)
 	if err != nil {
 		return nil, fmt.Errorf("failed to create containerd container: %w", err)
 	}
 	defer func() {
-		if retErr != nil {
+		if retErr != nil && shouldDelContainer {
```

Comment:
We should keep the container for other cleanup failures as well.

Reply:
ok, I find we can keep the container when following cases fail. Is it too much change? To reduce risk, maybe let's just do it for network cleanup?

1. Cleanup sandbox root directory
2. Cleanup volatile sandbox root directory
3. Cleanup sandbox files
4. **Cleanup network**
5. Delete sandbox task

Reply:
SG to limit the change to network for now.